File: atcoder/arc/arc043/b.py
mod = 10 ** 9 + 7
MX = 10 ** 5 + 5
N = int(raw_input())
dp = [[0] * MX for i in xrange(4)]
for i in xrange(N):
    d = int(raw_input())
    dp[0][d] += 1
    dp[1][d] += 1
    dp[2][d] += 1
    dp[3][d] += 1

for i in xrange(1, 5):
    for j in xrange(1, MX):
        dp[i - 1][j] = (dp[i - 1][j] + dp[i - 1][j - 1]) % mod
    if i == 4: break
    for j in xrange(1, MX):
        dp[i][j] = dp[i][j] * dp[i - 1][j / 2]

print dp[3][-1]

File: atcoder/arc/arc043/b.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

#define REP(i, a, b) for (int i = (int)(a); i < (int)(b); i++)
#define rep(i, a) REP(i, 0, a)
#define EACH(i, a) for (auto i: a)
#define ITR(x, a) for (__typeof(a.begin()) x = a.begin(); x != a.end(); x++)
#define ALL(a) (a.begin()), (a.end())

const ll MOD = 1000000007;
const int MX = 100005;

int N;
ll dp[5][MX];

int main(void) {
  cin >> N;
  rep(i, N) {
    int d;
    cin >> d;
    dp[0][d] = dp[1][d] = dp[2][d] = ++dp[3][d];
  }

  REP(i, 1, 5) {
    REP(j, 1, MX) dp[i - 1][j] = (dp[i - 1][j] + dp[i - 1][j - 1]) % MOD;
    REP(j, 1, MX) dp[i][j] = (dp[i][j] * dp[i - 1][j / 2]) % MOD;
  }

  cout << dp[3][MX - 1] << endl;
    
  return 0;
}
